### PR TITLE
Enable subdimensions on time column (type/Time)

### LINKED ISSUE #17598: Cannot group by `TIME` columns, because binning menu is not available and defaults to "Month"
**Describe the bug**
When trying to group by a `TIME` column, then no granularity options are shown and selecting the field will default to "Month", which fails during query.

Feels like a frontend issue with the incorrect default and no options shown.

**To Reproduce**
1. Custom question > (some table with a `TIME` column)
2. Try to group by the field - no binning options are available and it defaults to "Month"
image
3. Visualize will error with `Unsupported temporal bucketing: You can't bucket a :type/Time Field by :month.`

<details><summary>Full stacktrace</summary>

```
2021-08-25 19:56:46,522 ERROR middleware.catch-exceptions :: Error processing query: null
{:database_id 8,
 :started_at #t "2021-08-25T19:56:46.049213+02:00[Europe/Copenhagen]",
 :error_type :invalid-query,
 :json_query
 {:database 8,
  :query {:source-table 124, :aggregation [["count"]], :breakout [["field" 834 {:temporal-unit "month"}]]},
  :type "query",
  :parameters [],
  :middleware {:js-int-to-string? true, :add-default-userland-constraints? true}},
 :native nil,
 :status :failed,
 :class clojure.lang.ExceptionInfo,
 :stacktrace
 ["--> query_processor.middleware.validate_temporal_bucketing$validate_temporal_bucketing_STAR_.invokeStatic(validate_temporal_bucketing.clj:36)"
  "query_processor.middleware.validate_temporal_bucketing$validate_temporal_bucketing_STAR_.invoke(validate_temporal_bucketing.clj:29)"
  "query_processor.middleware.validate_temporal_bucketing$validate_temporal_bucketing$fn__50072.invoke(validate_temporal_bucketing.clj:49)"
  "query_processor.middleware.auto_parse_filter_values$auto_parse_filter_values$fn__45579.invoke(auto_parse_filter_values.clj:43)"
  "query_processor.middleware.wrap_value_literals$wrap_value_literals$fn__41713.invoke(wrap_value_literals.clj:161)"
  "query_processor.middleware.annotate$add_column_info$fn__41588.invoke(annotate.clj:608)"
  "query_processor.middleware.permissions$check_query_permissions$fn__46880.invoke(permissions.clj:81)"
  "query_processor.middleware.pre_alias_aggregations$pre_alias_aggregations$fn__49001.invoke(pre_alias_aggregations.clj:40)"
  "query_processor.middleware.cumulative_aggregations$handle_cumulative_aggregations$fn__47081.invoke(cumulative_aggregations.clj:60)"
  "query_processor.middleware.resolve_joined_fields$resolve_joined_fields$fn__49298.invoke(resolve_joined_fields.clj:102)"
  "query_processor.middleware.resolve_joins$resolve_joins$fn__49611.invoke(resolve_joins.clj:171)"
  "query_processor.middleware.add_implicit_joins$add_implicit_joins$fn__45155.invoke(add_implicit_joins.clj:190)"
  "query_processor.middleware.large_int_id$convert_id_to_string$fn__47844.invoke(large_int_id.clj:59)"
  "query_processor.middleware.format_rows$format_rows$fn__47825.invoke(format_rows.clj:74)"
  "query_processor.middleware.add_default_temporal_unit$add_default_temporal_unit$fn__44449.invoke(add_default_temporal_unit.clj:23)"
  "query_processor.middleware.desugar$desugar$fn__47147.invoke(desugar.clj:21)"
  "query_processor.middleware.binning$update_binning_strategy$fn__45966.invoke(binning.clj:229)"
  "query_processor.middleware.resolve_fields$resolve_fields$fn__46683.invoke(resolve_fields.clj:34)"
  "query_processor.middleware.add_dimension_projections$add_remapping$fn__44804.invoke(add_dimension_projections.clj:314)"
  "query_processor.middleware.add_implicit_clauses$add_implicit_clauses$fn__45033.invoke(add_implicit_clauses.clj:147)"
  "query_processor.middleware.upgrade_field_literals$upgrade_field_literals$fn__50021.invoke(upgrade_field_literals.clj:40)"
  "query_processor.middleware.add_source_metadata$add_source_metadata_for_source_queries$fn__45318.invoke(add_source_metadata.clj:123)"
  "query_processor.middleware.reconcile_breakout_and_order_by_bucketing$reconcile_breakout_and_order_by_bucketing$fn__49173.invoke(reconcile_breakout_and_order_by_bucketing.clj:100)"
  "query_processor.middleware.auto_bucket_datetimes$auto_bucket_datetimes$fn__45526.invoke(auto_bucket_datetimes.clj:147)"
  "query_processor.middleware.resolve_source_table$resolve_source_tables$fn__46730.invoke(resolve_source_table.clj:45)"
  "query_processor.middleware.parameters$substitute_parameters$fn__48983.invoke(parameters.clj:111)"
  "query_processor.middleware.resolve_referenced$resolve_referenced_card_resources$fn__46782.invoke(resolve_referenced.clj:79)"
  "query_processor.middleware.expand_macros$expand_macros$fn__47531.invoke(expand_macros.clj:184)"
  "query_processor.middleware.add_timezone_info$add_timezone_info$fn__45327.invoke(add_timezone_info.clj:15)"
  "query_processor.middleware.splice_params_in_response$splice_params_in_response$fn__49974.invoke(splice_params_in_response.clj:32)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49184$fn__49188.invoke(resolve_database_and_driver.clj:31)"
  "driver$do_with_driver.invokeStatic(driver.clj:60)"
  "driver$do_with_driver.invoke(driver.clj:56)"
  "query_processor.middleware.resolve_database_and_driver$resolve_database_and_driver$fn__49184.invoke(resolve_database_and_driver.clj:25)"
  "query_processor.middleware.fetch_source_query$resolve_card_id_source_tables$fn__47771.invoke(fetch_source_query.clj:274)"
  "query_processor.middleware.store$initialize_store$fn__49983$fn__49984.invoke(store.clj:11)"
  "query_processor.store$do_with_store.invokeStatic(store.clj:44)"
  "query_processor.store$do_with_store.invoke(store.clj:38)"
  "query_processor.middleware.store$initialize_store$fn__49983.invoke(store.clj:10)"
  "query_processor.middleware.validate$validate_query$fn__50028.invoke(validate.clj:10)"
  "query_processor.middleware.normalize_query$normalize$fn__47907.invoke(normalize_query.clj:22)"
  "query_processor.middleware.add_rows_truncated$add_rows_truncated$fn__45173.invoke(add_rows_truncated.clj:35)"
  "query_processor.middleware.results_metadata$record_and_return_metadata_BANG_$fn__49959.invoke(results_metadata.clj:147)"
  "query_processor.middleware.constraints$add_default_userland_constraints$fn__47024.invoke(constraints.clj:42)"
  "query_processor.middleware.process_userland_query$process_userland_query$fn__49070.invoke(process_userland_query.clj:134)"
  "query_processor.middleware.catch_exceptions$catch_exceptions$fn__46964.invoke(catch_exceptions.clj:173)"
  "query_processor.reducible$async_qp$qp_STAR___38252$thunk__38253.invoke(reducible.clj:103)"
  "query_processor.reducible$async_qp$qp_STAR___38252.invoke(reducible.clj:109)"
  "query_processor.reducible$sync_qp$qp_STAR___38261$fn__38264.invoke(reducible.clj:135)"
  "query_processor.reducible$sync_qp$qp_STAR___38261.invoke(reducible.clj:134)"
  "query_processor$process_userland_query.invokeStatic(query_processor.clj:241)"
  "query_processor$process_userland_query.doInvoke(query_processor.clj:237)"
  "query_processor$fn__50118$process_query_and_save_execution_BANG___50127$fn__50130.invoke(query_processor.clj:253)"
  "query_processor$fn__50118$process_query_and_save_execution_BANG___50127.invoke(query_processor.clj:245)"
  "query_processor$fn__50162$process_query_and_save_with_max_results_constraints_BANG___50171$fn__50174.invoke(query_processor.clj:265)"
  "query_processor$fn__50162$process_query_and_save_with_max_results_constraints_BANG___50171.invoke(query_processor.clj:258)"
  "api.dataset$run_query_async$fn__56409.invoke(dataset.clj:56)"
  "query_processor.streaming$streaming_response_STAR_$fn__56388$fn__56389.invoke(streaming.clj:72)"
  "query_processor.streaming$streaming_response_STAR_$fn__56388.invoke(streaming.clj:71)"
  "async.streaming_response$do_f_STAR_.invokeStatic(streaming_response.clj:65)"
  "async.streaming_response$do_f_STAR_.invoke(streaming_response.clj:63)"
  "async.streaming_response$do_f_async$fn__16088.invoke(streaming_response.clj:84)"],
 :context :ad-hoc,
 :error "Unsupported temporal bucketing: You can't bucket a :type/Time Field by :month.",
 :row_count 0,
 :running_time 0,
 :preprocessed nil,
 :ex-data
 {:type :invalid-query,
  :field [:field 834 {:temporal-unit :month}],
  :base-type :type/Time,
  :unit :month,
  :valid-units #{:hour :second :default :hour-of-day :millisecond :minute-of-hour :minute}},
 :data {:rows [], :cols []}}

2021-08-25 19:56:46,558 DEBUG middleware.log :: POST /api/dataset 202 [ASYNC: completed] 512.7 ms (8 DB calls) App DB connections: 0/10 Jetty threads: 3/50 (4 idle, 0 queued) (59 total active threads) Queries in flight: 0 (0 queued)
```

</details>

Similar happens, when trying to bin a `DATE` column (e.g People.BirthDate) by "Hour"/"Minute":
`Unsupported temporal bucketing: You can't bucket a :type/Date Field by :hour.`

**Information about your Metabase Installation:**
Tested 0.37.9 thru 0.41.5

:arrow_down: Please click the :+1: reaction instead of leaving a `+1` or `update?` comment

### LINKED ISSUE #20882: Subdimensions/binnings do not exist for time column
For a column of the `time` type, there is no subdimensions. This is unlike `timestamp`.

**To Reproduce**

First create a table in e.g. PostgreSQL that contains some columns with `time`, e.g.

```
CREATE TABLE test.test (
	"time" time NULL,
	"timestamp" timestamp NULL,
	count integer NULL
);
```

and the populate it with some data.

1. New, Question
2. Choose the above database and table
3. Summarize, Number of distinct values of
4. Highlight `Timestamp` and we got several choices of subdimensions (correctly, as expected)

image

5. **However**, try to choose `Time` and there is **no way** to choose a subdimension

image

**Expected behavior**

In Step 5, there should be a choice of dimensions for `Time` column, just like it is for Step 4 (for `Timestamp` column).

### PR DESCRIPTION
**Pull request by ariya**
To verify manually:

First create a table in e.g. PostgreSQL that contains some columns with `time`, e.g.

```
CREATE TABLE test.test (
	"time" time NULL,
	"timestamp" timestamp NULL,
	count integer NULL
);
```

and the populate it with some data.

1. New, Question
2. Choose the above database and table
3. Summarize, Number of distinct values of
4. Highlight `Timestamp` and we got several choices of subdimensions (correctly, as expected)

image

5. **However**, try to highlight `Time` and observe.

**Before this PR**

There is no subdimension for `Time`. This is confusing because it seems that things are working just fine for `Timestamp`.

image

**After this PR**

The behavior for `Time` matches that of `Timestamp`, i.e. there are suitable dimensions for it.

image

### CODE REVIEW

**Review by flamber — APPROVED** [review]
This seems to work fine. Limiting of options can be handled in followup PRs.

Would love to understand the reason for the limitation initially - perhaps there's something that I'm overlooking - or perhaps it's just like the sorting limitation?

**Comment by ariya** [comment]
> Would love to understand the reason for the limitation initially - perhaps there's something that I'm overlooking - or perhaps it's just like the sorting limitation?

The limitation was added years ago. I can only postulate that many things were not working well with temporal fields back then, hence the constraint.

**Comment by camsaul** [comment]
We originally didn't support temporal bucketing on `TIME` columns, but that was maybe five years ago. So I'm guessing that's why we don't show options there. That being said you definitely shouldn't be showing date-specific options like `day`, `week`, `month`, etc. for a `TIME` column since it makes no sense and will probably break

File: test/metabase/api/table_test.clj [hunk 1]
```diff
@@ -713,7 +713,7 @@
         (mt/test-drivers (mt/normal-drivers-except #{:sparksql :mongo :oracle :redshift})
           (mt/dataset test-data-with-time
             (let [response (mt/user-http-request :rasta :get 200 (format "table/%d/query_metadata" (mt/id :users)))]
-              (is (= []
+              (is (= @#'table-api/datetime-dimension-indexes
                      (dimension-options-for-field response "last_login_time"))))))))))
```
Comment (camsaul) [inline test/metabase/api/table_test.clj:717]: This isn't right we shouldn't be showing **date**time options for a `TIME`

**Review by camsaul — CHANGES_REQUESTED** [review]
We shouldn't be showing **date**time options for a `TIME`... `week(my_time_column)` makes no sense for example. I think we need to rework a bit more of the code in this Table API namespace.. we could build `dimension-options` a little more elegantly. I can take a look at that

File: test/metabase/api/table_test.clj (hunk 1 above)
Comment (ariya) [reply test/metabase/api/table_test.clj:717]: Good points! @flamber also gave me some additional ideas so I'm incorporating all into the refinement.

**Comment by flamber** [comment]
@ariya "Hour of Day" would be the same as "Hour", since there's no day, so can probably be hidden. It still defaults to Month:
image

**Comment by ariya** [comment]
> @ariya "Hour of Day" would be the same as "Hour", since there's no day, so can probably be hidden. It still defaults to Month:

Great catch, thank you! This should be fixed now:

image

**Comment by dpsutton** [comment]
I'm following the repo:

```sql
testing=# CREATE TABLE time_test (
        "time" time NULL,
        "timestamp" timestamp NULL,
        count integer NULL
);
CREATE TABLE
testing=# insert into time_test (time, timestamp, count) values ('12:45', now(), 3), ('4:42', now(), 7);
INSERT 0 2
testing=# select * from time_test;
   time   |         timestamp          | count
----------+----------------------------+-------
 12:45:00 | 2022-03-15 12:47:30.878597 |     3
 04:42:00 | 2022-03-15 12:47:30.878597 |     7
(2 rows)

testing=#
```

image

image

File: src/metabase/api/table.clj [hunk 2]
```diff
@@ -132,7 +133,7 @@
                      :type "type/DateTime"})
                   ;; note the order of these options corresponds to the order they will be shown to the user in the UI
                   [[(deferred-tru "Minute") "minute"]
-                   [(deferred-tru "Hour") "hour"]
+                   [hour-str "hour"]
```
Comment (camsaul) [inline src/metabase/api/table.clj:136]: 👍 on doing this for `Hour` but why not do it for `Minute` also

**Review by camsaul — APPROVED** [review]
I'm cool with these changes so we show the options in the FE but can you open a follow-on issue with the bug Dan mentioned? If someone tries to use the new options it will fail so we should make sure we address that too

File: src/metabase/api/table.clj (hunk 2 above)
Comment (ariya) [reply src/metabase/api/table.clj]: Good catch, thanks! I've added that.

**Comment by ariya** [comment]
> I'm cool with these changes so we show the options in the FE but can you open a follow-on issue with the bug Dan mentioned? If someone tries to use the new options it will fail so we should make sure we address that too

Done: it's now tracked in #21269.

**Pull request merged**